blockchain/interceptors.go

```diff
@@ -16,29 +28,65 @@ func (p Processor) jobValidationInterceptor(srv interface{}, ss grpc.ServerStrea
 		return status.Errorf(codes.InvalidArgument, "missing metadata")
 	}
 
-	jobAddressMd, ok := md[JobAddressHeader]
-	if !ok {
-		return status.Errorf(codes.InvalidArgument, "missing snet-job-address")
+	paymentHandler, err := p.newPaymentHandlerByType(md)
+	if err != nil {
+		return err
 	}
 
-	jobAddressBytes := common.FromHex(jobAddressMd[0])
+	err = paymentHandler.validatePayment()
+	if err != nil {
+		return err
+	}
 
-	jobSignatureMd, ok := md[JobSignatureHeader]
-	if !ok {
-		return status.Errorf(codes.InvalidArgument, "missing snet-job-signature")
+	err = handler(srv, ss)
+
+	return paymentHandler.completePayment(err)
+}
+
+func (p Processor) newPaymentHandlerByType(md metadata.MD) (paymentHandlerType, error) {
+	paymentTypeMd, ok := md[PaymentTypeHeader]
+
+	paymentType := JobPaymentType
+	if ok && len(paymentTypeMd) > 0 {
+		paymentType = paymentTypeMd[0]
 	}
 
-	jobSignatureBytes := common.FromHex(jobSignatureMd[0])
+	switch {
+	case paymentType == JobPaymentType:
+		return newJobPaymentHandler(&p, md), nil
+	case paymentType == EscrowPaymentType:
+		return newEscrowPaymentHandler(), nil
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unexpected \"%v\", value: \"%v\"", PaymentTypeHeader, paymentType)
+	}
+}
 
-	if p.IsValidJobInvocation(jobAddressBytes, jobSignatureBytes) {
-		err := handler(srv, ss)
-		if err == nil {
-			p.CompleteJob(jobAddressBytes, jobSignatureBytes)
-		}
-		return err
+type paymentHandlerType interface {
+	validatePayment() error
+	completePayment(error) error
+}
+
+func getBigInt(md metadata.MD, key string) (*big.Int, error) {
+	array := md.Get(key)
+	if len(array) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "missing \"%v\"", key)
 	}
 
-	return status.Errorf(codes.Unauthenticated, "job invocation not valid")
+	value := big.NewInt(0)
+	err := value.UnmarshalText([]byte(array[0]))
```

Comment:
[hexutil.DecodeBig ](https://godoc.org/github.com/ethereum/go-ethereum/common/hexutil#DecodeBig) allows to decode a hex string to a big.Int
hexutil allows to decode a string to byte array as well.

Reply:
Thanks, it seems that using ```-bin``` as key postfix means that value is binary and will be passed in ```base64``` format. And this convention is supported by all gRPC clients. I will test it and will use it for signature field.

Regarding this method it should decode string which contains decimal number. This format is used for incrementing ids such as channel_id and channel_nonce. It is not used in this code and not tested yet. I am going to test it and use it in next PR, so I would live it here if you don't mind.